wiki/Concept-Starting-and-orchestrating-Bots.md

```diff
@@ -22,17 +25,65 @@ I would recommend an infrastructure that would look like the following diagram:
 
 My recommendation would be to not take care of the orchestration in-house, but rather let external services take care of that (for now). I can see the benefits, which would especially help the management level through statistics and extended information, but I do not think that the orchestration should be of priority because of the rabbit hole that we might fall into security-wise.
+
+# Communication with ark automate dektop app
+
+I looked at several possible technologies/concepts that be applied in our case.
+
+1. Polling and Long-Polling
+2. Websockets
+3. Server-sent events
+4. Higher Abstraction Frameworks
+
+### Polling and Long-Polling
+
+Polling is an approach where the desktop app would permenantly 'ask' the there is a new robot to execute. Although it is quite easy to implement, the performance is bad.
+Long Polling aims to solve the performance issue by just 'asking' one time and waiting for an answer. The desktop app would 'ask' the server again after a fixed periode of time if the server does not have any new robots to execute. In principle a great approach but it lacks support of frameworks one could use and implementing it from scretch would be too time consuming
+
+### Websockets
+
+When using websockets we would open a bidirectional communication channel between the server side and the client side.Thus a great solution because real time communication is possible. Still, it is time consuming to implement and a successful connection is not always garantued as either element (proxy, firewall, …) prevent WebSocket connections between the client and the server or the browser does not support it yet.
+
+### Server-sent events
+
+This is an approach based on the concept of publish-subscribe. The client (desktop app) would subscribe to a 'channel' and the server would push new events (robots to be executed) on that channel. One of the main problem with SSE is the max of 6 client connections from a single host. Also, it is (subjectively) not the best approach to implement with node.js.
+
+### Higher Abstraction Frameworks
+
+There are some frameworks using one (or more) of these basic approaches but are one abstraction layer higher for the user. In particular, I looked at pusher and socket.io.
+Pusher is based on a publish subscribe approach simular to SSE and it seems to be very easy to use. The biggest disadvantage was that it is not open-source but ultimatly one would have to pay for it. Socket.io is a highly popular (> 4 Mio. weekly npm downloads) open source framework. It is mainly based on websocket communication but is also able to seamlessly switch to a long polling approach in the background if websockets do not work.
+
+### Decision
+
+In the end, I decided for socket.io as it is open-cource, well-supported, 'easy' to use, robust and websocket based enabling a bidirectional communication.
```

Comment:
passive again or you could also refer to we, because by merging this we will all have agreed on that

Reply:
✅